core/src/stack/mod.rs

```diff
@@ -16,3 +16,6 @@ pub use outputs::StackOutputs;
 /// The number of stack registers which can be accessed by the VM directly. This is also the
 /// minimum stack depth enforced by the VM.
 pub const STACK_TOP_SIZE: usize = 16;
+
+/// Maximum number of elements allowed for the input and output stack.
+pub const STACK_DEPTH: usize = 16;
```

Comment:
Probably we should leave only one constant.

Reply:
Agreed; I would maybe have just `STACK_MAX_DEPTH`? And then document that 

- it's the maximum number of non-zero elements allowed on the stack at the beginning and end of the execution
- Anything more than that goes in the overflow table

Reply:
I think `MAX_STACK_DEPTH` may be a bit confusing because we also have the concept of stack depth defined as 16 + number of elements in the overflow table (e.g., accessible via `sdepth` instruction).

We could use `MIN_STACK_DEPTH` (as stack depth never drops below 16) - but I'd probably prefer to keep `STACK_TOP_SIZE` as I think `MIN_STACK_DEPTH` would be a bit confusing as well.

Reply:
I agree that `MAX_STACK_DEPTH` isn't ideal, but I also don't understand the naming behind "stack top size". In my mind, the "stack top" is the element on the top of the stack (e.g. [in C++](https://cplusplus.com/reference/stack/stack/top/)). Then "stack top size" is the size of that element.

Maybe `VISIBLE_STACK_DEPTH`, where "visible" refers to the fact that it's directly in the trace? If not, I find `MIN_STACK_DEPTH` to be less confusing than `STACK_TOP_SIZE`

Reply:
I tried to use "stack top" as a stand-alone concept that refers to the top 16 elements of the stack - but probably that's a bit confusing. Basically, would be good to have something like that because it is tied to a few things:

1. Only the top 16 elements can be initialize at the start of execution and remain populated at the end of execution.
2. Only the top 16 elements can be accessed directly via instructions.
3. Only the top 16 elements remain visible to the callee when we switch context via `call` or `syscall` instructions.
4. The depth of the stack never drops below 16 elements.

And saying "the top 16 elements of the stack" is a bit of a mouthful.

Given the options though - I probably would prefer `MIN_STACK_DEPTH` to `VISIBLE_STACK_DEPTH`.

Reply:
I think `MIN_STACK_DEPTH` with a long-ish docstring that explains what you just explained is a great solution.

Reply:
Oh, I just noticed this was added [here](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/core/src/stack/mod.rs#L16-L23); I think it's better if we replace `STACK_TOP_SIZE` with this definition (and docstring)

Reply:
@plafer Do you mean to replace the `STACK_TOP_SIZE` with `MIN_STACK_DEPTH`, but also leave the `STACK_DEPTH` constant as it was?

Reply:
No, I'm thinking just 1 constant (`STACK_TOP_SIZE`) renamed to `MIN_STACK_DEPTH`. As we discussed, we're going to describe what "stack top size" was trying to describe instead with "minimum stack depth"

Reply:
But in that case I already done this: I left only one constant (`MIN_STACK_DEPTH`) and created a docstring with a description based on the Bobbin's comment. Sorry, probably I'm missing something.

Reply:
We still have the constant `STACK_TOP_SIZE` [here](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/air/src/trace/stack/mod.rs#L13), right? My point is:

1. Rename `STACK_TOP_SIZE` to `MIN_STACK_DEPTH`
2. Add [this docstring](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/core/src/stack/mod.rs#L16-L22) to it
3. Remove [this constant](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/core/src/stack/mod.rs#L23)

Reply:
Ah, I see, there are two `STACK_TOP_SIZE` constants there! Yep, now I get it, I'll rework it as you said.

Reply:
Ah yes good point! I didn't even noticed it was defined a few times (`STACK_TOP_SIZE` in core, in air and `MIN_STACK_DEPTH`).

As you said I would define it just once in `core` [here](https://github.com/0xPolygonMiden/miden-vm/blob/b192c61216d4dacfe19cb8609ba25f6b0d74193d/core/src/stack/mod.rs#L18) (although renamed to `MIN_STACK_DEPTH`) and use that in all crates.

Reply:
One thing to note, we still use the term "stack top" pretty consistently throughout the codebase to refer to the top 16 elements of the stack (e.g., [here](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/core/src/lib.rs#L131) and [here](https://github.com/0xPolygonMiden/miden-vm/blob/2c23e12b60460828bbaaf4630f28ec212828b4df/assembly/src/ast/instruction/debug.rs#L12)). I wonder to what extent we can replace these with `StackInputs` and `StackOutputs` structs. Though, updating the debug instruction should probably be in a separate PR